File: packages/postgres-memory/src/bulk.ts
import { eq, inArray, sql } from "drizzle-orm";
import { drizzle } from "drizzle-orm/node-postgres";
import type { Pool } from "pg";
import { memories, rawMemories } from "./schema.js";

export type MemoryType = "profile" | "preference" | "context";

export interface RawMemoryRecord {
  id: string;
  tenantId: string;
  sessionId: string;
  runId: string;
  content: string;
  createdAt: Date;
}

export interface MemoryEntry {
  id: string;
  tenantId: string;
  type: MemoryType;
  title: string;
  content: string;
  createdAt: Date;
  updatedAt: Date;
  lastConfirmedAt: Date;
}

export interface NewMemory {
  type: MemoryType;
  title: string;
  content: string;
}

export interface MemoryUpdate {
  id: string;
  content: string;
}

/** 提取阶段（Phase 1）写入一条原始观察。跟 set_config 事务包装是因为这张表也在 RLS 保护下——
 * 见 scripts/migrations/2026-09-20-add-memories-rls.sql（Task 3）。 */
export async function insertRawMemory(
  pool: Pool,
  params: { tenantId: string; sessionId: string; runId: string; content: string },
): Promise<void> {
  const db = drizzle(pool);
  await db.transaction(async (tx) => {
    await tx.execute(sql`select set_config('app.tenant_id', ${params.tenantId}, true)`);
    await tx.insert(rawMemories).values({
      id: crypto.randomUUID(),
      tenantId: params.tenantId,
      sessionId: params.sessionId,
      runId: params.runId,
      content: params.content,
    });
  });
}

/** 找出"有未处理原始记忆"的全部租户。不能直接对 raw_memories 表做一次跨租户 DISTINCT 查询——
 * 那张表本身受 RLS 保护，在 opentalos_app 角色下任何没设置 app.tenant_id 的查询永远只返回空结果
 * （这是 RLS 该有的行为，不是 bug）。调用方（apps/worker 的定期整合循环，见 Task 7）先从
 * packages/postgres-tenancy 的 TenantStore.listTenants()（tenants 表明确排除在 RLS 范围之外）拿到
 * 全部租户 id 列表，传进来逐个检查。租户规模不大时这个 N+1 查询完全够用；如果未来租户规模变大，
 * 需要换一个不依赖逐租户扫描的机制。 */
export async function listTenantsWithUnconsolidatedRawMemories(pool: Pool, allTenantIds: string[]): Promise<string[]> {
  const db = drizzle(pool);
  const result: string[] = [];
  for (const tenantId of allTenantIds) {
    const hasWork = await db.transaction(async (tx) => {
      await tx.execute(sql`select set_config('app.tenant_id', ${tenantId}, true)`);
      const rows = await tx.select({ id: rawMemories.id }).from(rawMemories).where(eq(rawMemories.tenantId, tenantId)).limit(1);
      return rows.length > 0;
    });
    if (hasWork) result.push(tenantId);
  }
  return result;
}

export async function listRawMemoriesForTenant(pool: Pool, tenantId: string): Promise<RawMemoryRecord[]> {
  const db = drizzle(pool);
  return db.transaction(async (tx) => {
    await tx.execute(sql`select set_config('app.tenant_id', ${tenantId}, true)`);
    return tx.select().from(rawMemories).where(eq(rawMemories.tenantId, tenantId));
  });
}

export async function deleteRawMemories(pool: Pool, tenantId: string, ids: string[]): Promise<void> {
  if (ids.length === 0) return;
  const db = drizzle(pool);
  await db.transaction(async (tx) => {
    await tx.execute(sql`select set_config('app.tenant_id', ${tenantId}, true)`);
    await tx.delete(rawMemories).where(inArray(rawMemories.id, ids));
  });
}

export async function listMemoriesForTenant(pool: Pool, tenantId: string): Promise<MemoryEntry[]> {
  const db = drizzle(pool);
  return db.transaction(async (tx) => {
    await tx.execute(sql`select set_config('app.tenant_id', ${tenantId}, true)`);
    const rows = await tx.select().from(memories).where(eq(memories.tenantId, tenantId));
    return rows.map((row) => ({ ...row, type: row.type as MemoryType }));
  });
}

/** 整合阶段（Phase 2）的批量新增/更新入口。newEntries 是全新记忆（分配新 id），updates 是对
 * 已有记忆的改写（只更新 content，同时刷新 updatedAt/lastConfirmedAt）。 */
export async function upsertMemories(
  pool: Pool,
  tenantId: string,
  changes: { newEntries: NewMemory[]; updates: MemoryUpdate[] },
): Promise<void> {
  const db = drizzle(pool);
  await db.transaction(async (tx) => {
    await tx.execute(sql`select set_config('app.tenant_id', ${tenantId}, true)`);
    for (const entry of changes.newEntries) {
      await tx.insert(memories).values({ id: crypto.randomUUID(), tenantId, type: entry.type, title: entry.title, content: entry.content });
    }
    for (const update of changes.updates) {
      await tx
        .update(memories)
        .set({ content: update.content, updatedAt: new Date(), lastConfirmedAt: new Date() })
        .where(eq(memories.id, update.id));
    }
  });
}

export async function deleteMemories(pool: Pool, tenantId: string, ids: string[]): Promise<void> {
  if (ids.length === 0) return;
  const db = drizzle(pool);
  await db.transaction(async (tx) => {
    await tx.execute(sql`select set_config('app.tenant_id', ${tenantId}, true)`);
    await tx.delete(memories).where(inArray(memories.id, ids));
  });
}

File: services/memory-service/db/bulk.py
import asyncio
import uuid
from dataclasses import dataclass
from datetime import datetime
from typing import Literal

from psycopg_pool import AsyncConnectionPool

MemoryType = Literal["profile", "preference", "context"]


@dataclass
class RawMemoryRecord:
    id: str
    tenant_id: str
    session_id: str
    run_id: str
    content: str
    created_at: datetime


@dataclass
class MemoryEntry:
    id: str
    tenant_id: str
    type: MemoryType
    title: str
    content: str
    created_at: datetime
    updated_at: datetime
    last_confirmed_at: datetime


@dataclass
class NewMemory:
    type: MemoryType
    title: str
    content: str


@dataclass
class MemoryUpdate:
    id: str
    content: str


async def insert_raw_memory(
    pool: AsyncConnectionPool, tenant_id: str, session_id: str, run_id: str, content: str
) -> None:
    async with pool.connection() as conn, conn.transaction():
        await conn.execute("select set_config('app.tenant_id', %s, true)", (tenant_id,))
        await conn.execute(
            "INSERT INTO raw_memories (id, tenant_id, session_id, run_id, content) VALUES (%s, %s, %s, %s, %s)",
            (str(uuid.uuid4()), tenant_id, session_id, run_id, content),
        )


async def list_tenants_with_pending(pool: AsyncConnectionPool, all_tenant_ids: list[str]) -> list[str]:
    """找出"有未处理原始记忆"的全部租户。不能直接对 raw_memories 表做一次跨租户查询——那张表受
    RLS 保护，在 opentalos_app 角色下任何没设置 app.tenant_id 的查询永远只返回空结果。调用方
    （apps/worker 的定期整合循环）先从 tenants 表拿到全部租户 id，传进来逐个检查（并行）。"""

    async def check(tenant_id: str) -> str | None:
        async with pool.connection() as conn, conn.transaction():
            await conn.execute("select set_config('app.tenant_id', %s, true)", (tenant_id,))
            cur = await conn.execute("SELECT id FROM raw_memories WHERE tenant_id = %s LIMIT 1", (tenant_id,))
            rows = await cur.fetchall()
            return tenant_id if rows else None

    results = await asyncio.gather(*(check(tenant_id) for tenant_id in all_tenant_ids))
    return [tenant_id for tenant_id in results if tenant_id is not None]


async def list_raw_memories_for_tenant(pool: AsyncConnectionPool, tenant_id: str) -> list[RawMemoryRecord]:
    async with pool.connection() as conn, conn.transaction():
        await conn.execute("select set_config('app.tenant_id', %s, true)", (tenant_id,))
        cur = await conn.execute(
            "SELECT id, tenant_id, session_id, run_id, content, created_at FROM raw_memories WHERE tenant_id = %s",
            (tenant_id,),
        )
        rows = await cur.fetchall()
        return [
            RawMemoryRecord(id=r[0], tenant_id=r[1], session_id=r[2], run_id=r[3], content=r[4], created_at=r[5])
            for r in rows
        ]


async def delete_raw_memories(pool: AsyncConnectionPool, tenant_id: str, ids: list[str]) -> None:
    if not ids:
        return
    async with pool.connection() as conn, conn.transaction():
        await conn.execute("select set_config('app.tenant_id', %s, true)", (tenant_id,))
        await conn.execute("DELETE FROM raw_memories WHERE tenant_id = %s AND id = ANY(%s)", (tenant_id, ids))


async def list_memories_for_tenant(pool: AsyncConnectionPool, tenant_id: str) -> list[MemoryEntry]:
    async with pool.connection() as conn, conn.transaction():
        await conn.execute("select set_config('app.tenant_id', %s, true)", (tenant_id,))
        cur = await conn.execute(
            "SELECT id, tenant_id, type, title, content, created_at, updated_at, last_confirmed_at "
            "FROM memories WHERE tenant_id = %s",
            (tenant_id,),
        )
        rows = await cur.fetchall()
        return [
            MemoryEntry(
                id=r[0], tenant_id=r[1], type=r[2], title=r[3], content=r[4],
                created_at=r[5], updated_at=r[6], last_confirmed_at=r[7],
            )
            for r in rows
        ]


async def upsert_memories(
    pool: AsyncConnectionPool, tenant_id: str, new_entries: list[NewMemory], updates: list[MemoryUpdate]
) -> None:
    """整合阶段的批量新增/更新入口。故意不用 ON CONFLICT（跟 TS 版本一致）——同一个租户的整合
    循环在这个试点里始终是单 worker 串行处理，不会有并发写同一个 title 的情况；如果未来有多个
    worker 实例并发跑整合，需要重新评估。"""
    async with pool.connection() as conn, conn.transaction():
        await conn.execute("select set_config('app.tenant_id', %s, true)", (tenant_id,))
        for entry in new_entries:
            await conn.execute(
                "INSERT INTO memories (id, tenant_id, type, title, content) VALUES (%s, %s, %s, %s, %s)",
                (str(uuid.uuid4()), tenant_id, entry.type, entry.title, entry.content),
            )
        for update in updates:
            await conn.execute(
                "UPDATE memories SET content = %s, updated_at = now(), last_confirmed_at = now() "
                "WHERE tenant_id = %s AND id = %s",
                (update.content, tenant_id, update.id),
            )


async def delete_memories(pool: AsyncConnectionPool, tenant_id: str, ids: list[str]) -> None:
    if not ids:
        return
    async with pool.connection() as conn, conn.transaction():
        await conn.execute("select set_config('app.tenant_id', %s, true)", (tenant_id,))
        await conn.execute("DELETE FROM memories WHERE tenant_id = %s AND id = ANY(%s)", (tenant_id, ids))


async def search_memories(pool: AsyncConnectionPool, tenant_id: str, query: str) -> list[dict[str, object]]:
    async with pool.connection() as conn, conn.transaction():
        await conn.execute("select set_config('app.tenant_id', %s, true)", (tenant_id,))
        pattern = f"%{query}%"
        cur = await conn.execute(
            "SELECT title, content FROM memories WHERE tenant_id = %s AND (title ILIKE %s OR content ILIKE %s)",
            (tenant_id, pattern, pattern),
        )
        rows = await cur.fetchall()
        return [{"key": r[0], "value": r[1], "score": 0.0} for r in rows]
